src/Simplex_tree/include/gudhi/Simplex_tree.h

```diff
@@ -873,19 +898,26 @@ class Simplex_tree {
     GUDHI_CHECK(*vi != null_vertex(), "cannot use the dummy null_vertex() as a real vertex");
     res_insert = curr_sib->members_.emplace(*vi, Node(curr_sib, filtration));
     if (!res_insert.second) {
-      // if already in the complex
-      if (res_insert.first->second.filtration() > filtration) {
+      if constexpr (Options::store_filtration){
+        // if already in the complex
+        auto oldFil = res_insert.first->second.filtration();
+        unify_births(res_insert.first->second.filtration_raw(), filtration);
+        if (oldFil == res_insert.first->second.filtration()) {
```

Comment:
For multi, this copies a vector. Should we make `unify_births` return a bool saying if things changed? (does that come almost for free?)

Reply:
Should be possible, as `add_point` in the multi-critical filtration could also return a bool to indicate if a point was actually added or not.